File: go/code/46.permutations.go
package main

func permute(nums []int) [][]int {
	ans := [][]int{}

	var backtrack func(path []int, used []bool)
	backtrack = func(path []int, used []bool) {
		if len(path) == len(nums) {
			temp := make([]int, len(path))
			copy(temp, path)
			ans = append(ans, temp)
			return
		}

		for i := 0; i < len(nums); i++ {
			if used[i] {
				continue
			}

			used[i] = true
			path = append(path, nums[i])
			backtrack(path, used)

			used[i] = false
			path = path[:len(path)-1]
		}
	}

	used := make([]bool, len(nums))
	backtrack([]int{}, used)
	return ans
}

File: ruby/46.permutations.rb
# frozen_string_literal: true

# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
  used = Array.new(nums.size, false)
  ans = []
  back_track([], used, nums, ans)
  ans
end

def back_track(path, used, nums, ans)
  if path.size == nums.size
    ans << path.dup
    return ans
  end

  nums.each_with_index do |num, index|
    next if used[index]

    used[index] = true
    path << num
    back_track(path, used, nums, ans)

    used[index] = false
    path.pop
  end
end
